man/rc-environ.8
.\" Copyright (c) 2026 The OpenRC Authors.
.\" See the Authors file at the top-level directory of this distribution and
.\" https://github.com/OpenRC/openrc/blob/HEAD/AUTHORS
.\"
.\" This file is part of OpenRC. It is subject to the license terms in
.\" the LICENSE file found in the top-level directory of this
.\" distribution and at https://github.com/OpenRC/openrc/blob/HEAD/LICENSE
.\" This file may not be copied, modified, propagated, or distributed
.\"    except according to the terms contained in the LICENSE file.
.\"
.Dd Sep 18, 2026 
.Dt RC-ENVIRON 8 SMM
.Os OpenRC
.Sh NAME
.Nm rc-environ
.Nd Print the stored enviroment variables for a given service
.Sh SYNOPSIS
.Nm
.Op -s <service>
.Op -r <runlevel>
.Op -en0
.Sh DESCRIPTION
.Nm
prints the stored environment variables for a service.
.Pp
The options are as follows:
.Bl -tag -width ".Fl test , test string"
.It Fl e , -export
Prepend "export" to printed variables
.It Fl n , -no-escape
Do not perform shell sensitive escape on variable values
.It Fl 0 , -null
End each output line with NUL, not newline
.It Fl s , -service Ar <arg>
Add service to the list of environments to print
.It Fl r , -runlevel Ar <arg>
Add services in runlevel to the list of environments to print
.It Fl h , -help
Display this help output
.It Fl C , -nocolor
Disable color output
.It Fl V, -version
Display software version
.It Fl v , -verbose
Run verbosely
.It Fl q , -quiet
Run quietly (repeat to suppress errors)
.It Fl U , -user
Run in user mode
.El
.Sh AUTHORS
.An Matthias Geiger <werdahias@debian.org>
.An navi <navi@vlhl.dev>


NEWS
OpenRC NEWS

This file will contain a list of notable changes for each release.

OpenRC-0.13.2
=============

A chroot variable has been added to the service script variables.
This fixes the support for running a service in a chroot.
This is documented in man 8 openrc-run.

The netmount service now mounts nfs file systems.
This change was made to correct a fix for an earlier bug.

OpenRC-0.13
===========

/sbin/rc was renamed to /sbin/openrc and /sbin/runscript was renamed to
/sbin/openrc-run due to naming conflicts with other software.

Backward compatible symbolic links are currently in place so your
system will keep working if you are using the old names; however, it is
strongly advised that you migrate to the new names because the symbolic
links will be removed in the future.
Warnings have been added to assist with this migration; however, due to the
level of noise they produce, they only appear in verbose mode in this release.

The devfs script now handles the initial mounting and setup of the
/dev directory. If /dev has already been mounted by the kernel or an
initramfs, devfs will remount /dev with the correct mount options
instead of mounting a second /dev over the existing mount point.

It attempts to mount /dev from fstab first if an entry exists there. If
it doesn't it attempts to mount devtmpfs if it is configured in the
kernel. If not, it attempts to mount tmpfs.
If none of these is available, an error message is displayed and static
/dev is assumed.

OpenRC-0.12
===========

The net.* scripts, originally from Gentoo Linux, have
been removed. If you need these scripts, look for a package called
netifrc, which is maintained by them.


src/rc-environ/rc-environ.c
#include <einfo.h>
#include <errno.h>
#include <getopt.h>
#include <rc.h>
#include <queue.h>
#include <string.h>

#include "_usage.h"
#include "helpers.h"

const char *applet;
const char *extraopts;
const char *usagestring = "rc-environ [-s <service> ...] [-r <runlevel> ...] [-en0]\n";
const char getoptstring[] = "ens:r:0" getoptstring_COMMON;
const struct option longopts[] = {
	{ "export",          no_argument, NULL, 'e' },
	{ "no-escape",       no_argument, NULL, 'n' },
	{ "null",            no_argument, NULL, '0' },
	{ "service",   required_argument, NULL, 's' },
	{ "runlevel",  required_argument, NULL, 'r' },
	longopts_COMMON
};

const char *const longopts_help[] = {
	"Prepend \"export\" to printed variables",
	"Do not perform shell sensitive escape on variable values",
	"End each output line with NUL, not newline",
	"Add service to the list of environments to print",
	"Add services in runlevel to the list of environments to print",
	longopts_help_COMMON
};

static int
set_environment(const char *service, char *vars[])
{
	RC_DEPTREE *deptree = _rc_deptree_load(0, NULL);
	RC_STRINGLIST *reexports = rc_deptree_depend(deptree, service, "reexport");
	RC_STRING *reexport;

	if (!(rc_service_state(service) & RC_SERVICE_STARTING)) {
		eerror("service %s is not in the starting state", service);
		return 1;
	}

	for (const char *var; (var = *vars); vars++) {
		size_t name_len = strcspn(var, "=");

		TAILQ_FOREACH(reexport, reexports, entries) {
			if (strncmp(reexport->value, var, name_len) != 0)
				continue;
			if (var[name_len] == '=')
				rc_service_putenv(service, var);
			else
				rc_service_setenv(service, var, getenv(var));
			goto next;
		}

		ewarn("service %s does not reexport %.*s, skipping", service, (int)name_len, var);
next:;
	}

	rc_stringlist_free(reexports);
	rc_deptree_free(deptree);
	return 0;
}

static void escape_variable(const char *value) {
	char ch;

	printf("$'");
	for (const char *p = value; *p; p++) {
		switch ((ch = *p)) {
		case '\a': ch = 'a'; break;
		case '\b': ch = 'b'; break;
		case '\f': ch = 'f'; break;
		case '\n': ch = 'n'; break;
		case '\r': ch = 'r'; break;
		case '\t': ch = 't'; break;
		case '\v': ch = 'v'; break;
		/* \e is non-standard. */
		case 0x1b: ch = 'e'; break;
		case '"':
		case '\'':
		case '\\':
			break;
		default:
			putchar(ch);
			continue;
		}
		printf("\\%c", ch);
	}
	printf("\'");
}

static int
print_environment(RC_STRINGLIST *services, bool escape, bool export, char sep)
{
	/* from POSIX.2024 Shell Command Language 2.2, excluding `=` */
	static const char special_chars[] = "|&;<>()$`\\\"' \t\n*?[]^-!#~%{},";
	RC_DEPTREE *deptree = _rc_deptree_load(0, NULL);
	RC_STRING *service;
	char *value = NULL;
	int ret = 0;

	if (!services)
		services = rc_services_in_state(RC_SERVICE_STARTED);

	TAILQ_FOREACH(service, services, entries) {
		RC_STRINGLIST *reexports = rc_deptree_depend(deptree, service->value, "reexport");
		RC_STRING *reexport;

		TAILQ_FOREACH(reexport, reexports, entries) {
			if (rc_service_getenv(service->value, reexport->value, &value) == -1) {
				ewarn("%s: failed to read variable '%s' for '%s': %s",
					applet, reexport->value, service->value, strerror(errno));
				ret = 1;
				continue;
			}

			if (export)
				fputs("export ", stdout);
			printf("%s=", reexport->value);
			if (escape && strpbrk(value, special_chars))
				escape_variable(value);
			else
				fputs(value, stdout);
			fputc(sep, stdout);
		}

		rc_stringlist_free(reexports);
	}

	free(value);
	rc_deptree_free(deptree);
	return ret;
}

int main(int argc, char **argv) {
	RC_STRINGLIST *services = NULL, *in_runlevel;
	bool export = false, escape = true;
	char sep = '\n';

	for (int opt; (opt = getopt_long(argc, argv, getoptstring, longopts, NULL)) != -1;) {
		switch (opt) {
		case 'e': export = true; break;
		case 'n': escape = false; break;
		case '0': sep = '\0'; break;
		case 'r':
			if (!services)
				services = rc_stringlist_new();
			in_runlevel = rc_services_in_runlevel(optarg);
			TAILQ_CONCAT(services, in_runlevel, entries);
			rc_stringlist_free(in_runlevel);
			break;
		case 's':
			if (!services)
				services = rc_stringlist_new();
			rc_stringlist_addu(services, optarg);
			break;
		case_RC_COMMON_GETOPT
		}
	}

	argc -= optind;
	argv += optind;

	if (!argc)
		return print_environment(services, escape, export, sep);

	return set_environment(argv[0], &argv[1]);
}
